File: python/xl-client/dial_xl/field_groups.py
from typing import Iterator

from dial_xl.events import ObservableObserver, notify_observer, Event
from dial_xl.field import Field
from dial_xl.reader import _Reader
from dial_xl.utils import _validate_index


class FieldGroup(ObservableObserver):
    __before: str = "  "
    __fields: list[Field]
    __separator = " = "
    __formula: str | None
    __after: str = "\n"
    __decorator_indices: dict[str, int]
    __field_indices: dict[str, int]

    def __init__(self, formula: str | None):
        self.__formula = formula
        self.__fields = []
        self.__field_indices = {}

    def to_dsl(self):
        """Converts the field to DSL format."""

        if not self.__fields:
            raise ValueError("List of fields is empty")

        return (
            f"{self.__before}"
            f"{''.join(f.to_dsl() for f in self.__fields)}"
            f"{'' if self.__formula is None else self.__separator + self.__formula}"
            f"{self.__after}"
        )

    @property
    def formula(self) -> str | None:
        """Get the formula of the field group."""
        return self.__formula

    @formula.setter
    @notify_observer
    def formula(self, value: str | None):
        """Set the formula of the field group, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        self.__formula = value

    def get_field(self, name: str) -> Field:
        """Get a field by name."""
        index = self.__find_field(name)
        if index == -1:
            raise ValueError(f"Field '{name}' not found")

        return self.__fields[index]

    def has_field(self, name: str) -> bool:
        """Check if a field exists."""
        return self.__find_field(name) != -1

    @notify_observer
    def add_field(self, field: Field):
        """Add a field to the group, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        if field.name in self.__field_indices:
            raise ValueError(f"Field '{field.name}' already exists")

        field._attach(self)
        field_count = len(self.__fields)
        if field_count > 0:
            field._set_before(", ")
        self.__field_indices[field.name] = field_count
        self.__fields.append(field)

        return field

    @notify_observer
    def insert_field(self, index: int, field: Field):
        """Insert a field at a specified index, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        _validate_index(index, len(self.__fields), "Field")

        field._attach(self)

        if index == 0:
            self.__fields[index]._set_before(", ")
        elif index > 0:
            field._set_before(", ")
        self.__fields.insert(index, field)
        self.__update_field_indices()

    @notify_observer
    def remove_field(self, name: str) -> Field:
        """Remove a field by name, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        index = self.__find_field(name)
        if index == -1:
            raise ValueError(f"Field '{name}' not found")

        field = self.__fields.pop(index)
        field._detach()
        field._set_before("")
        if index < len(self.__fields):
            self.__fields[index]._set_before("")
        self.__update_field_indices()

        return field

    def __find_field(self, name: str) -> int:
        return self.__field_indices.get(name, -1)

    def __update_field_indices(self):
        self.__field_indices = {
            field.name: index for index, field in enumerate(self.__fields)
        }

    @property
    def field_names(self) -> Iterator[str]:
        """Enumerate field names."""
        return (field.name for field in self.fields)

    @property
    def fields(self) -> Iterator[Field]:
        """Enumerate fields."""
        return self.__fields.__iter__()

    @property
    def field_count(self) -> int:
        """Get the number of fields."""
        return len(self.__fields)

    def _notify_before(self, event: Event):
        if self._observer:
            self._observer._notify_before(event)

        sender = event.sender
        if isinstance(sender, Field):
            match event.method_name:
                case "name":
                    self.__on_field_rename(sender.name, event.kwargs["value"])

    def __on_field_rename(self, old_name: str, new_name: str):
        index = self.__find_field(old_name)
        if index == -1:
            raise ValueError(f"Field '{old_name}' not found")

        if new_name in self.__field_indices:
            raise ValueError(f"Field '{new_name}' already exists")

        self.__field_indices[new_name] = self.__field_indices.pop(old_name)

    @classmethod
    def from_field(cls, field: Field, formula: str | None):
        """Create a FieldGroup from a field and formula."""
        result = cls(formula)
        result.add_field(field)
        return result

    @classmethod
    def _deserialize(cls, reader: _Reader) -> "FieldGroup":
        result = cls(None)
        result.__before = reader.next(lambda d: d["span"]["from"])

        for field_entity in reader.entity.get("fields", []):
            field_reader = reader.with_entity(field_entity)
            field = Field._deserialize(field_reader)
            result.__fields.append(field)
            field._attach(result)
            reader.position = field_reader.position

        formula = reader.entity.get("formula")
        if formula:
            result.__separator = reader.next(formula["span"]["from"])
            result.__formula = reader.next(formula["span"]["to"])

        result.__after = reader.till_linebreak()
        result.__update_field_indices()

        return result


class FieldGroups(ObservableObserver):
    __field_groups: list[FieldGroup | str]
    __field_group_indices: list[int]

    def __init__(self):
        self.__field_groups = []
        self.__field_group_indices = []

    def to_dsl(self) -> str:
        """Convert the field groups to DSL format."""
        return "".join(
            f if isinstance(f, str) else f.to_dsl() for f in self.__field_groups
        )

    def __len__(self):
        """Get the number of field groups."""
        return len(self.__field_group_indices)

    def __iter__(self) -> Iterator[FieldGroup]:
        """Iterate over field groups."""
        return (self.__field_groups[index] for index in self.__field_group_indices)

    def __getitem__(self, index: int) -> FieldGroup:
        """Get a field group by index."""
        index = self.__to_group_index(index)
        return self.__field_groups[index]

    @notify_observer
    def __setitem__(self, index: int, value: FieldGroup):
        """Set a field group at a specified index, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        index = self.__to_group_index(index)
        value._attach(self)
        old_value = self.__field_groups[index]
        self.__field_groups[index] = value
        old_value._detach()

    @notify_observer
    def __delitem__(self, index: int):
        """Remove a field group from the table, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        self.__remove_group(index)

    @notify_observer
    def append(self, value: FieldGroup):
        """Add a field group to the table, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        value._attach(self)
        self.__field_group_indices.append(len(self.__field_groups))
        self.__field_groups.append(value)

    @notify_observer
    def insert(self, index: int, value: FieldGroup):
        """Insert a field group at a specified index, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        index = self.__to_group_index(index)
        value._attach(self)
        self.__field_groups.insert(index, value)
        self.__update_field_group_indices()

    @notify_observer
    def pop(self, index: int) -> FieldGroup:
        """Remove a field group from the table, which invalidates all compilation and computation results, as well as any sheet parsing errors."""
        return self.__remove_group(index)

    def __remove_group(self, index: int):
        index = self.__to_group_index(index)
        field_group = self.__field_groups.pop(index)
        field_group._detach()
        self.__update_field_group_indices()

        return field_group

    def __to_group_index(self, index: int) -> int:
        if index < 0 or index >= len(self.__field_group_indices):
            raise ValueError(f"The field group number {index} does not exist.")

        return self.__field_group_indices[index]

    def __update_field_group_indices(self):
        self.__field_group_indices = [
            index
            for index, group in enumerate(self.__field_groups)
            if isinstance(group, FieldGroup)
        ]

    @classmethod
    def _deserialize(cls, reader: _Reader) -> "FieldGroups":
        result = cls()
        for field_entity in reader.entity.get("fields", []):
            field_reader = reader.with_entity(field_entity)
            unparsed = field_reader.next_unparsed(lambda d: d["span"]["from"])
            if unparsed:
                result.__field_groups.append(unparsed + field_reader.till_linebreak())
            field_group = FieldGroup._deserialize(field_reader)
            result.__field_groups.append(field_group)
            field_group._attach(result)
            reader.position = field_reader.position

        result.__update_field_group_indices()

        return result

File: frontend/libs/parser/src/lib/EditDslApi/Field/FieldGroups.ts
import { notifyObserver, ObservableObserver, Reader } from '../utils';
import { FieldGroup } from './FieldGroup';

/**
 * A container class for multiple FieldGroup objects,
 * Each FieldGroup is tracked by an internal array and an index map.
 */
export class FieldGroups extends ObservableObserver {
  /**
   * The underlying storage of items, which may be a mix of string snippets or FieldGroup instances.
   */
  private _fieldGroups: Array<FieldGroup | string> = [];

  /**
   * Tracks the indexes of actual FieldGroup objects within `_fieldGroups`.
   */
  private _fieldGroupIndices: number[] = [];

  /**
   * Constructs an empty `FieldGroups` container.
   */
  constructor() {
    super();
  }

  /**
   * Converts the field groups to a DSL.
   *
   * @returns The combined DSL representation of all items in `_fieldGroups`.
   */
  public toDSL(): string {
    return this._fieldGroups
      .map((item) => (typeof item === 'string' ? item : item.toDSL()))
      .join('');
  }

  /**
   * Returns how many actual FieldGroup objects are stored.
   *
   * @returns The number of `FieldGroup` items.
   */
  public get length(): number {
    return this._fieldGroupIndices.length;
  }

  /**
   * Provides an iterator over all FieldGroup items in the order they appear,
   * skipping any string entries.
   */
  public [Symbol.iterator](): IterableIterator<FieldGroup> {
    const self = this;
    let current = 0;

    return {
      next(): IteratorResult<FieldGroup> {
        if (current < self._fieldGroupIndices.length) {
          const arrIndex = self._fieldGroupIndices[current++];
          const group = self._fieldGroups[arrIndex] as FieldGroup;

          return { value: group, done: false };
        }

        return { value: undefined as unknown as FieldGroup, done: true };
      },
      [Symbol.iterator](): IterableIterator<FieldGroup> {
        return this;
      },
    };
  }

  /**
   * Retrieves a FieldGroup by zero-based index in the list of groups,
   * throwing if out of range.
   *
   * @param index - The zero-based index among actual field groups (not raw strings).
   * @throws {Error} If the index is out of bounds.
   * @returns The matching FieldGroup.
   */
  public getItem(index: number): FieldGroup {
    index = this.toGroupIndex(index);

    return this._fieldGroups[index] as FieldGroup;
  }

  /**
   * Replaces the FieldGroup at the given index, detaching the old group
   * and attaching the new one. Triggers observer notifications.
   *
   * @param index - The zero-based index among field groups.
   * @param value - The new FieldGroup to set.
   * @throws {Error} If the index is out of bounds.
   */
  @notifyObserver()
  public setItem(index: number, value: FieldGroup): void {
    index = this.toGroupIndex(index);
    value.attach(this);

    const actualIdx = this._fieldGroupIndices[index];
    const oldValue = this._fieldGroups[actualIdx] as FieldGroup;
    this._fieldGroups[actualIdx] = value;
    oldValue.detach();
  }

  /**
   * Removes the FieldGroup at the specified index, detaching it.
   * Triggers observer notifications.
   *
   * @param index - The zero-based index among field groups.
   * @returns The removed FieldGroup.
   * @throws {Error} If the index is out of bounds.
   */
  @notifyObserver()
  public deleteItem(index: number): FieldGroup {
    return this.removeGroup(index);
  }

  /**
   * Appends a new FieldGroup at the end of `_fieldGroups`. Triggers observer notifications.
   *
   * @param value - The FieldGroup to add.
   */
  @notifyObserver()
  public append(value: FieldGroup): void {
    value.attach(this);
    this._fieldGroupIndices.push(this._fieldGroups.length);
    this._fieldGroups.push(value);
  }

  /**
   * Inserts a new FieldGroup before the group at the specified index, pushing it forward.
   * Triggers observer notifications.
   *
   * @param index - The zero-based index among field groups where to insert.
   * @param value - The FieldGroup to insert.
   * @throws {Error} If the index is out of bounds.
   */
  @notifyObserver()
  public insert(index: number, value: FieldGroup): void {
    index = this.toGroupIndex(index);
    value.attach(this);

    this._fieldGroups.splice(index, 0, value);

    this.updateFieldGroupIndices();
  }

  /**
   * Removes and returns a FieldGroup by index.
   * Triggers observer notifications.
   *
   * @param index - The zero-based index among field groups.
   * @returns The removed FieldGroup.
   * @throws {Error} If the index is out of bounds.
   */
  @notifyObserver()
  public pop(index: number): FieldGroup {
    return this.removeGroup(index);
  }

  /**
   * Internally removes a FieldGroup at the given index from `_fieldGroups`,
   * detaching it and re-building `_fieldGroupIndices`.
   *
   * @param index - The zero-based index among field groups.
   * @returns The removed FieldGroup.
   * @throws {Error} If the index is out of bounds.
   */
  private removeGroup(index: number): FieldGroup {
    index = this.toGroupIndex(index);
    const fieldGroup = this._fieldGroups.splice(index, 1)[0] as FieldGroup;
    fieldGroup.detach();
    this.updateFieldGroupIndices();

    return fieldGroup;
  }

  /**
   * Ensures the given index is valid for the current list of field groups.
   *
   * @param index - The zero-based index of the group to check.
   * @throws {Error} If the index is out of range.
   */
  public toGroupIndex(index: number): number {
    if (index < 0 || index >= this._fieldGroupIndices.length) {
      throw new Error(`The field group number ${index} does not exist.`);
    }

    return this._fieldGroupIndices[index];
  }

  /**
   * Re-computes `_fieldGroupIndices` by scanning `_fieldGroups` for actual FieldGroup items.
   * This should be called whenever we add, remove, or reorder field groups in `_fieldGroups`.
   */
  private updateFieldGroupIndices(): void {
    this._fieldGroupIndices = [];
    this._fieldGroups.forEach((item, idx) => {
      if (item instanceof FieldGroup) {
        this._fieldGroupIndices.push(idx);
      }
    });
  }

  /**
   * Deserializes a `FieldGroups` object by reading field group definitions (and raw snippets)
   * from the given reader. Each group is appended to `result`.
   *
   * @param reader - A `Reader` positioned where field definitions begin in the DSL.
   * @returns A fully constructed `FieldGroups` instance with all field groups parsed.
   */
  public static deserialize(reader: Reader): FieldGroups {
    const result = new FieldGroups();

    const fieldsData = reader.entity?.fields ?? [];
    for (const fieldEntity of fieldsData) {
      const fieldReader = reader.withEntity(fieldEntity);

      // Possibly read an unparsed snippet
      const unparsed = fieldReader.nextUnparsed((d) => d.span.from);
      if (unparsed) {
        result._fieldGroups.push(unparsed + fieldReader.tillLinebreak());
      }

      // Now parse an actual FieldGroup
      const fieldGroup = FieldGroup.deserialize(fieldReader);
      result._fieldGroups.push(fieldGroup);
      fieldGroup.attach(result);

      reader.position = fieldReader.position;
    }

    result.updateFieldGroupIndices();

    return result;
  }
}
